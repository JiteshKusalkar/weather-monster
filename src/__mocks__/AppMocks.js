export const validProps = {
  cities: [
    {
      id: 1283240,
      name: 'Kathmandu',
      main: {
        temp: 57.2,
        temp_max: 57.2,
        temp_min: 57.2
      },
      weather: [
        {
          description: 'mist',
          icon: '50d',
          id: 701,
          main: 'Mist'
        }
      ]
    }
  ],
  deleteCity: jest.fn(),
  fetchCurrentWeatherById: jest.fn()
};

export const emptyProps = {
  cities: [],
  deleteCity: () => null,
  fetchCurrentWeatherById: () => null
};
