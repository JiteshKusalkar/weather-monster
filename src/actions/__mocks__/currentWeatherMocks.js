import { FETCH_CURRENT_WEATHER_REQUEST } from "../currentWeather";

export const response = [
  { type: FETCH_CURRENT_WEATHER_REQUEST },
  {
    payload: {
      base: 'stations',
      clouds: { all: 75 },
      cod: 200,
      coord: { lat: 27.72, lon: 85.32 },
      dt: 1576927921,
      id: 1283240,
      main: {
        feels_like: 12.66,
        humidity: 62,
        pressure: 1016,
        temp: 15,
        temp_max: 15,
        temp_min: 15
      },
      name: 'Kathmandu',
      sys: {
        country: 'NP',
        id: 9201,
        sunrise: 1576890297,
        sunset: 1576927681,
        type: 1
      },
      timezone: 20700,
      visibility: 7000,
      weather: [
        { description: 'broken clouds', icon: '04n', id: 803, main: 'Clouds' }
      ],
      wind: { deg: 270, speed: 2.6 }
    },
    type: 'ADD_CITY'
  }
];
