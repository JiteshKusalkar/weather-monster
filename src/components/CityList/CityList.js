import styled from 'styled-components';
import React from 'react';

const unit = process.env.REACT_APP_UNITS === 'imperial' ? 'C' : 'F';

const CityList = ({ cities, onDelete }) => {
  const renderWeatherCard = city => {
    const {
      name,
      main: { temp, temp_min: min, temp_max: max },
      weather
    } = city;

    return (
      <WeatherCard key={city.id} onClick={onDelete(city.id)}>
        <CityName>{name}</CityName>
        <TemperatureSummary>
          {weather.map(elem => (
            <WeatherIcon key={`icon-${name}-${elem.id}`} icon={elem.icon} />
          ))}
          <CurrentTemperature>{temp}</CurrentTemperature>
          <Unit size='lg'>{unit}</Unit>
        </TemperatureSummary>
        <Divider />
        <Temperature>
          <Label>Min.</Label>
          <Value>{min}</Value>
          <Unit size='sm'>{unit}</Unit>
        </Temperature>
        <Temperature>
          <Label>Max.</Label>
          <Value>{max}</Value>
          <Unit size='sm'>{unit}</Unit>
        </Temperature>
      </WeatherCard>
    );
  };

  return cities && <Wrapper>{cities.map(renderWeatherCard)}</Wrapper>;
};

export default CityList;

const Wrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
  margin: 30px;
`;

const WeatherCard = styled.div`
  box-shadow: 0 6px 7px 0 rgba(0, 151, 182, 0.3);
  padding: 20px;
  background-color: #0097b6;
  color: #ffffff;
  width: 100%;
  margin: 0 20px 20px 0;
  border-radius: 5px;
  cursor: pointer;

  @media (min-width: 1024px) {
    width: calc((100% / 3) - 60px);
  }
`;

const TemperatureSummary = styled.div`
  display: flex;
  justify-content: center;
`;

const CurrentTemperature = styled.div`
  font-size: 60px;
  margin: 0;

  &:after {
    content: '°';
  }
`;

const Unit = styled.span`
  font-size: ${props => `${props.size === 'lg' ? '40px' : '20px'}`};
  margin: 0;
  vertical-align: top;
`;

const WeatherIcon = styled.div`
  width: 45px;
  height: 45px;
  background: ${props => `
    url(http://openweathermap.org/img/w/${props.icon}.png)
    repeat scroll 0% 0%
    transparent`};
`;

const CityName = styled.div`
  padding: 10px;
  text-align: center;
  font-size: 20px;
`;

const Divider = styled.div`
  width: 30px;
  border: 2px solid #ffffff;
  margin: 20px auto;
`;

const Temperature = styled.div`
  width: 50%;
  display: inline-block;
  text-align: center;
`;

const Value = styled.span`
  font-size: 30px;
  margin: 0;

  &:after {
    content: '°';
  }
`;

const Label = styled.div``;
