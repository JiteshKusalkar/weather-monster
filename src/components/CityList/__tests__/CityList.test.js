import { shallow } from 'enzyme';
import 'jest-styled-components';
import React from 'react';
import renderer from 'react-test-renderer';
import unroll from 'unroll';

import { emptyProps, validProps } from '../__mocks__/CityListMock';
import { Unit, WeatherIcon, WeatherCard } from '../CityList';
import CityList from '../';

const shallowInstance = props => shallow(<CityList {...props} />);
unroll.use(it);

describe('Snapshot: CityList', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowInstance(emptyProps);
  });

  unroll(
    'should match the snapshot for #propType',
    (done, args) => {
      wrapper.setProps(args.value);
      expect(wrapper).toMatchSnapshot();
      done();
    },
    [
      ['propType', 'value'],
      ['valid props', validProps],
      ['empty props', emptyProps]
    ]
  );
});

describe('Unit: CityList', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowInstance(validProps);
  });

  it('should delete city on click of weather card', () => {
    wrapper.find(WeatherCard).simulate('click');

    expect(validProps.onDelete).toHaveBeenCalledWith(1283240);
  });
});

unroll(
  'should have font size #fontSize for #size size',
  (done, args) => {
    const tree = renderer.create(<Unit size={args.size} />).toJSON();
    expect(tree).toHaveStyleRule('font-size', args.fontSize);
    done();
  },
  [['size', 'fontSize'], ['lg', '40px'], ['sm', '20px']]
);

it('should render the image corresponding to icon', () => {
  const tree = renderer.create(<WeatherIcon icon='50d' />).toJSON();
  expect(tree).toHaveStyleRule(
    'background',
    'url(http://openweathermap.org/img/w/50d.png) repeat scroll 0% 0% transparent'
  );
});
