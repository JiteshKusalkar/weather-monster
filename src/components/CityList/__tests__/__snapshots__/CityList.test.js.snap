// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot: CityList should match the snapshot for "empty props" 1`] = `ShallowWrapper {}`;

exports[`Snapshot: CityList should match the snapshot for "valid props" 1`] = `ShallowWrapper {}`;
